### PR TITLE
rustbuild: Document many more parts of the build

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit expands the bootstrap build system's `README.md` as well as ensuring
that all API documentation is present and up-to-date. Additionally a new
`config.toml.example` file is checked in with commented out versions of all
possible configuration values.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I'm also more than willing to add any other documentation for rustbuild, so if anything is unclear or not documented well, please let me know!

File: src/bootstrap/build/channel.rs [hunk 1]
```diff
@@ -8,7 +8,13 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-use std::env;
```
Comment (steveklabnik) [inline src/bootstrap/build/channel.rs]: i'm guessing this just fixes a warning? no other code was modified in this file

File: src/bootstrap/build/channel.rs [hunk 2]
```diff
@@ -32,7 +41,8 @@ pub fn collect(build: &mut Build) {
         }
     }
 
-    // FIXME: this is duplicating makefile logic
```
Comment (steveklabnik) [inline src/bootstrap/build/channel.rs]: this fixme is just no longer relevant/

File: src/bootstrap/build/compile.rs [hunk 3]
```diff
@@ -8,6 +8,14 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+//! Implementation of compiling various phases of the compiler and standard
+//! library.
+//!
+//! This module contains some of the real meat in the rustbuild build syste
```
Comment (steveklabnik) [inline src/bootstrap/build/compile.rs]: typo: system

File: src/bootstrap/build/mod.rs [hunk 4]
```diff
@@ -414,16 +491,7 @@ impl Build {
...
-        // In stage0 we're using a previously released stable compiler, so we
-        // use the stage0 bootstrap key. Otherwise we use our own build's
-        // bootstrap key.
-        let bootstrap_key = if compiler.is_snapshot(self) {
-            &self.bootstrap_key_stage0
-        } else {
-            &self.bootstrap_key
-        };
-        cargo.env("RUSTC_BOOTSTRAP_KEY", bootstrap_key);
+        self.add_bootstrap_key(compiler, &mut cargo);
```
Comment (steveklabnik) [inline src/bootstrap/build/mod.rs]: this commit is described as docs only, so i just want to flag this small refactoring as a "you did want this in here, right?" I don't think it's a problem, just wanted to make sure it's intentional.

File: src/bootstrap/build/cc.rs [hunk 5]
```diff
@@ -8,6 +8,29 @@
...
+//! 6. "cc"
+//!
+//! Some of this logic is implemented here, but much of it is farmed out to the
+//! `gcc` crate itself, so we end up having the same fallbacks as there.
+//! Similar logic is then used to find a C++ compiler, just some s/cc/c++/ is
+//! used.
+//!
+//! It is intended that after this module has run that no C/C++ compiler will
+//! ever be probed for as part of the build, but rather the compiler's found
+//! here will be used for everything.
```
Comment (durka) [inline src/bootstrap/build/cc.rs]: This sentence has some questionable grammatical choices ("for" at the end of the clause, and apostrophe for pluralization).

File: src/bootstrap/build/channel.rs [hunk 6]
```diff
@@ -8,7 +8,13 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-use std::env;
+//! Build configuration for Rust's release channels.
+//!
+//! Implements the stable/beta/nightly channel distinctions by setting various
+//! flags like the `unstable_features`, calculating variables like `release` and
+//! `package_vers`, and otherwise indicating to the compiler what it should
+//! print out as part of it's version information.
```
Comment (durka) [inline src/bootstrap/build/channel.rs]: Another rogue apostrophe.

File: src/bootstrap/README.md [hunk 7]
```diff
@@ -42,6 +86,12 @@ build/
     debug/
     release/
 
+  # Output of the dist-related steps like dist-std, dist-rustc, and dist-docs
+  dist/
+
+  # Temporary directory used for various input/output sa part of various stages
```
Comment (brson) [inline src/bootstrap/README.md]: "sa"

File: src/bootstrap/README.md [hunk 8]
```diff
@@ -108,3 +172,66 @@ build/
...
+dive too much into the actual build system itself.
+
+After that, each module in rustbuild should have enough documentation to keep
+you up and running. Some general areas that you may be interested in modifying
+are:
+
+* Adding a new build tool? Take a look at `build/step.rs` for examples of other
+  tools, as well as `build/mod.rs`.
+* Adding a new crate? Look no further! Adding crates can be done by just
+  configuring all `Cargo.toml` files accordingly.
```
Comment (brson) [inline src/bootstrap/README.md]: This isn't clear to me. I'd guess that adding a crate means creating a new directory somewhere, writing a Cargo.toml file, and ensuring that it's mentioned in one other toml file. There's also the crates.io case. This document might say when it's appropriate to add a crates.io dependency.

**Comment by brson** [comment]
r=me

File: src/bootstrap/build/channel.rs (hunk 2 above)
Comment (alexcrichton) [inline src/bootstrap/build/channel.rs]: Ah yeah this is just gonna stick around until the makefiles go away, so no need to keep the FIXME

File: src/bootstrap/build/mod.rs (hunk 4 above)
Comment (alexcrichton) [inline src/bootstrap/build/mod.rs]: Ah yeah this was just a drive by fix, hopefully small though!

**Comment by alexcrichton** [comment]
@bors: retry

* #33435

**Pull request merged**